### PR TITLE
Reflection_Engine: Add support for new fragments attributes

### PR DESCRIPTION
**Pull request by adecler**
### NOTE: Depends on 
https://github.com/BHoM/BHoM/pull/1093

   
### Issues addressed by this PR
See https://github.com/BHoM/BHoM/pull/1093 for details

### CODE REVIEW

File: BHoM_Engine/Modify/AddFragment.cs [hunk 1]
```diff
@@ -41,13 +41,29 @@ public static partial class Modify
...
             IBHoMObject o = iBHoMObject.DeepClone();
 
-            if (!replace)
-                o.Fragments.Add(fragment);
+            // Give a warning if the fragment is supposed to be unique but is found on the object
+            List<Type> currentFragmentTypes = iBHoMObject.Fragments.Select(x => x.GetType()).ToList();
+            foreach (Type restriction in fragment.GetType().UniquenessRestrictions())
+            {
+                if (currentFragmentTypes.Any(x => restriction.IsAssignableFrom(x)))
+                    Engine.Reflection.Compute.RecordWarning("There is already a fragment of type " + restriction + " on this object. \nThe Fragment will still be aded but consider reviewing this task as fragments of that type are supposed to be unique.");
```
Comment (FraserGreenroyd) [inline BHoM_Engine/Modify/AddFragment.cs:53]:
```suggestion
                    Engine.Reflection.Compute.RecordWarning("There is already a fragment of type " + restriction + " on this object. \nThe Fragment will still be added but consider reviewing this task as fragments of that type are supposed to be unique.");
```

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]

**Comment by adecler** [comment]
This PR hasn't been reviewed nor merged in 3 weeks and is now past the deadline for feature PRs on this milestone. 
Since it doesn't seem to be as urgent as we initially thought, I will park this for now

**Review by IsakNaslundBh — APPROVED** [review]
Had to rebase and forcepush to resolve conflicts and get this to merge with master, hope you are ok with that @adecler .

I am generally happy with the functionality of this.

Would be good if we had some mechanism for replacing a unique attribute, but can come in subsequent PRs. This is, if I have IMaterialProperties and unique, and have a Steel on an object and try to add a Concrete. Would for this case be good if we had a mechanism for replacing the Steel with the new Concrete.

This can all be resolved later though, as this is a great first step!

Approving, but as I made the last commits by the force push, would be good if @FraserGreenroyd had a chance to look through.

**Comment by IsakNaslundBh** [comment]
@BHoM check compliance

**Comment by IsakNaslundBh** [comment]
@BHoMBot check compliance

**Comment by IsakNaslundBh** [comment]
@BHoMBot check compliance

File: BHoM_Engine/BHoM_Engine.csproj [hunk 2]
```diff
@@ -61,14 +61,16 @@
...
+    <Compile Include="Query\UniquenessRestrictions.cs" />
+    <Compile Include="Query\CanTarget.cs" />
     <Compile Include="Query\Hash.cs" />
     <Compile Include="Create\BHoMGroup.cs" />
     <Compile Include="Create\IObject\RandomObject.cs" />
     <Compile Include="Create\CustomObject.cs" />
     <Compile Include="Modify\RemoveFragment.cs" />
     <Compile Include="Modify\AddFragment.cs" />
+	<Compile Include="Modify\SetHashFragment.cs" />
+	<Compile Include="Objects\EqualityComparers\HashComparer.cs" />
```
Comment (FraserGreenroyd) [inline BHoM_Engine/BHoM_Engine.csproj:73]: Is it just me, or are these indented more?

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]
This looks good to me, just blocking till oM PR is ready to avoid accidental merge! Can be dismissed or changed to approval once oM is ready to go :smile:

**Comment by adecler** [comment]
@BHoMBot check copyright

**Comment by FraserGreenroyd** [comment]
@BHoMBot check code compliance

**Review by FraserGreenroyd — APPROVED** [review]
This LGTM now :smile:

**Pull request merged**